lahja/asyncio/endpoint.py

```diff
@@ -297,8 +300,8 @@ class AsyncioEndpoint(BaseEndpoint):
 
     _futures: Dict[RequestID, "asyncio.Future[BaseEvent]"]
 
-    _full_connections: Dict[str, RemoteEndpoint]
-    _half_connections: Set[RemoteEndpoint]
+    _full_connections: Dict[str, RemoteEndpointAPI]
+    _half_connections: Set[RemoteEndpointAPI]
```

Comment:
I'm inclined to push back but **very lightly**.  

you are correct that this would break spectacularly if a `TrioRemoteEndpoint` made its way in.  However, the benefit I see here is that this results in `mypy` enforcing that we (*mostly*) **only** interact with public APIs on this object.

I think the benefit outweighs the negatives since I think it's unlikely that the `trio/asyncio` mixup would go un-noticed since the two frameworks fail spectacularly when you try to run one from the other.

Reply:
Counter argument: But **when** you actually need to call an API on the `AsyncioRemoteEndpoint` (which is perfectly fine) you would be forced to do a `cast()` which is kind of ugly as it just means "stop it mypy, I know what this thing is and you should trust me".

Reply:
I agree that if we'd ever want to use a `cast` then we're dealing with `AsyncioRemoteEndpoint` and that's the type this should be declared as, but I also agree that if we want to force ourselves to stick to the public API this is a neat way to do so!

Something like this would do both at once, only providing access to the public API but also preventing Trio classes from being accepted in the wrong places:

- `RemoteEndpointAPI` (where the public API is defined)
  - `AsyncioRemoteEndpointAPI`  (intentionally kept empty, _half_connections would use this)
    - `AsyncioRemoteEndpoint` (the actual implementation) 
  - `TrioRemoteEndpointAPI`
    - `TrioRemoteEndpoint`

but that's probably overkill.